File: lib/clustering.ts
import * as tf from "@tensorflow/tfjs";
import { GitHubRepo } from "./github";

export interface ClusteredRepo {
  repo: GitHubRepo;
  clusterId: number;
  coordinates: [number, number];
}

export async function clusterRepositories(
  repos: GitHubRepo[]
): Promise<ClusteredRepo[]> {
  // Convert repos to feature vectors
  const features = tf.tensor2d(
    repos.map((repo) => [
      repo.stargazers_count,
      repo.forks_count,
      repo.open_issues_count,
      repo.size,
      repo.watchers_count,
      repo.topics.length,
      repo.language ? 1 : 0,
    ])
  );

  // Reduce dimensions and cluster
  const reduced = await reduceDimensions(features);
  const clustered = tf.tidy(() => {
    const coords = reduced.slice([0, 0], [-1, 2]) as tf.Tensor2D;
    const labels = reduced.slice([0, 2], [-1, 1]) as tf.Tensor2D;
    return { coords, labels };
  });

  // Map results to repositories
  const clusteredRepos = repos.map((repo, i) => ({
    repo,
    clusterId: clustered.labels.arraySync()[i][0],
    coordinates: clustered.coords.arraySync()[i] as [number, number],
  }));

  // Clean up tensors
  features.dispose();
  reduced.dispose();
  clustered.coords.dispose();
  clustered.labels.dispose();

  return clusteredRepos;
}

interface ClusterResult {
  labels: number[];
  corePoints: number[];
}

function dbscan(
  points: tf.Tensor2D,
  eps: number,
  minPts: number
): ClusterResult {
  const numPoints = points.shape[0];
  const labels = new Array(numPoints).fill(-1);
  const corePoints: number[] = [];
  let clusterId = 0;

  // Calculate pairwise distances
  const distances = tf.tidy(() => {
    const a = tf.sum(tf.square(points), 1, true);
    const b = tf.matMul(points, points.transpose());
    return tf.sqrt(tf.add(a, tf.transpose(a)).sub(tf.mul(b, 2))) as tf.Tensor2D;
  });

  // Find neighbors within eps distance
  const neighbors = [];
  for (let i = 0; i < numPoints; i++) {
    const pointNeighbors = [];
    for (let j = 0; j < numPoints; j++) {
      if ((distances.arraySync() as number[][])[i][j] <= eps) {
        pointNeighbors.push(j);
      }
    }
    neighbors.push(pointNeighbors);
  }

  // DBSCAN algorithm
  for (let i = 0; i < numPoints; i++) {
    if (labels[i] !== -1) continue;

    if (neighbors[i].length < minPts) {
      labels[i] = 0; // Noise
      continue;
    }

    clusterId++;
    labels[i] = clusterId;
    corePoints.push(i);

    const seedSet = new Set(neighbors[i]);
    seedSet.delete(i);

    while (seedSet.size > 0) {
      const current = seedSet.values().next().value;
      seedSet.delete(current);

      if (labels[current] === 0) {
        labels[current] = clusterId;
      }

      if (labels[current] !== -1) continue;

      labels[current] = clusterId;

      if (neighbors[current].length >= minPts) {
        neighbors[current].forEach((n) => seedSet.add(n));
      }
    }
  }

  distances.dispose();
  return { labels, corePoints };
}

async function reduceDimensions(features: tf.Tensor2D): Promise<tf.Tensor2D> {
  const numPoints = features.shape[0];

  // For very small datasets, use simpler scaling
  if (numPoints < 3) {
    return tf.tidy(() => {
      const normalized = tf
        .sub(features, tf.min(features))
        .div(tf.sub(tf.max(features), tf.min(features)).add(tf.scalar(1e-6)));
      return tf.mul(normalized, tf.scalar(2)).sub(tf.scalar(1)) as tf.Tensor2D;
    });
  }

  try {
    // Center and scale the data
    const mean = tf.mean(features, 0);
    const std = tf.sqrt(tf.mean(tf.square(tf.sub(features, mean)), 0));
    const normalized = tf.div(
      tf.sub(features, mean),
      std.add(tf.scalar(1e-6))
    ) as tf.Tensor2D;

    // Calculate pairwise distances
    const squaredDists = tf.tidy(() => {
      const a = tf.sum(tf.square(normalized), 1, true);
      const b = tf.matMul(normalized, normalized.transpose());
      return tf.add(a, tf.transpose(a)).sub(tf.mul(b, 2)) as tf.Tensor2D;
    });

    // Convert distances to probabilities (t-SNE)
    const sigma = tf.scalar(1.0);
    const probs = tf.tidy(() => {
      const negDists = tf.neg(tf.div(squaredDists, tf.mul(sigma, 2)));
      const expDists = tf.exp(negDists);
      const rowSums = tf.sum(expDists, 1, true);
      return tf.div(expDists, rowSums) as tf.Tensor2D;
    });

    // Initialize low-dimensional representation
    const outputDim = 2;
    let lowDim = tf
      .randomNormal([numPoints, outputDim])
      .mul(tf.scalar(0.0001)) as tf.Tensor2D;

    // Gradient descent
    const learningRate = tf.scalar(100);
    const numIterations = Math.min(
      100,
      Math.max(50, Math.floor(numPoints * 0.5))
    );

    for (let i = 0; i < numIterations; i++) {
      tf.tidy(() => {
        const lowDimSquaredDists = tf.tidy(() => {
          const a = tf.sum(tf.square(lowDim), 1, true);
          const b = tf.matMul(lowDim, lowDim.transpose());
          return tf.add(a, tf.transpose(a)).sub(tf.mul(b, 2)) as tf.Tensor2D;
        });

        const lowDimProbs = tf.tidy(() => {
          const negDists = tf.neg(lowDimSquaredDists);
          const expDists = tf.exp(negDists);
          const rowSums = tf.sum(expDists, 1, true);
          return tf.div(expDists, rowSums) as tf.Tensor2D;
        });

        const gradients = tf.tidy(() => {
          const diff = tf.sub(probs, lowDimProbs);
          return tf.matMul(diff, lowDim).mul(learningRate) as tf.Tensor2D;
        });

        lowDim = tf.add(lowDim, gradients) as tf.Tensor2D;
      });
    }

    // Clean up intermediate tensors
    squaredDists.dispose();
    probs.dispose();

    // Normalize final output and add cluster labels
    return tf.tidy(() => {
      const minVals = tf.min(lowDim, 0);
      const maxVals = tf.max(lowDim, 0);
      const normalized = tf.div(
        tf.sub(lowDim, minVals),
        tf.sub(maxVals, minVals).add(tf.scalar(1e-6))
      ) as tf.Tensor2D;

      // Apply DBSCAN clustering
      const { labels } = dbscan(normalized, 0.2, 5);

      // Add cluster labels as third dimension
      const labelsTensor = tf.tensor1d(labels).expandDims(1);
      const clustered = tf.concat([normalized, labelsTensor], 1) as tf.Tensor2D;

      lowDim.dispose();
      labelsTensor.dispose();
      return clustered;
    });
  } catch (error) {
    console.error("Error in dimensionality reduction:", error);
    // Fallback to simple 2D projection
    return tf.tidy(() => {
      const normalized = tf
        .sub(features, tf.min(features))
        .div(tf.sub(tf.max(features), tf.min(features)).add(tf.scalar(1e-6)));
      return tf.slice(normalized, [0, 0], [-1, 2]) as tf.Tensor2D;
    });
  }
}

File: backend/app/services/clustering.py
import numpy as np
import tensorflow as tf
from typing import List, Tuple
from ..models import GitHubRepo, ClusteredRepo

class ClusteringService:
    def __init__(self):
        self.eps = 0.2
        self.min_pts = 5

    def extract_features(self, repos: List[GitHubRepo]) -> np.ndarray:
        """Convert repositories to feature vectors."""
        features = []
        for repo in repos:
            features.append([
                repo.stargazers_count,
                repo.forks_count,
                repo.open_issues_count,
                repo.size,
                repo.watchers_count,
                len(repo.topics),
                1 if repo.language else 0
            ])
        return np.array(features, dtype=np.float32)

    def reduce_dimensions(self, features: np.ndarray) -> np.ndarray:
        """Reduce dimensionality using t-SNE like approach."""
        num_points = features.shape[0]
        
        # For very small datasets, use simpler scaling
        if num_points < 3:
            normalized = (features - np.min(features)) / (np.max(features) - np.min(features) + 1e-6)
            return normalized * 2 - 1

        # Center and scale the data
        mean = np.mean(features, axis=0)
        std = np.std(features, axis=0) + 1e-6
        normalized = (features - mean) / std

        # Calculate pairwise distances
        squared_dists = np.sum(normalized**2, axis=1)[:, np.newaxis] + \
                       np.sum(normalized**2, axis=1)[np.newaxis, :] - \
                       2 * np.dot(normalized, normalized.T)

        # Convert distances to probabilities
        sigma = 1.0
        probs = np.exp(-squared_dists / (2 * sigma**2))
        np.fill_diagonal(probs, 0)
        probs = probs / np.sum(probs, axis=1, keepdims=True)

        # Initialize low-dimensional representation
        output_dim = 2
        low_dim = np.random.normal(0, 0.0001, (num_points, output_dim))

        # Gradient descent
        learning_rate = 100
        num_iterations = min(100, max(50, int(num_points * 0.5)))

        for _ in range(num_iterations):
            # Calculate low-dimensional pairwise distances
            low_dim_dists = np.sum(low_dim**2, axis=1)[:, np.newaxis] + \
                           np.sum(low_dim**2, axis=1)[np.newaxis, :] - \
                           2 * np.dot(low_dim, low_dim.T)
            
            # Convert to probabilities
            low_dim_probs = np.exp(-low_dim_dists)
            np.fill_diagonal(low_dim_probs, 0)
            low_dim_probs = low_dim_probs / np.sum(low_dim_probs, axis=1, keepdims=True)

            # Calculate gradients
            gradients = np.dot(probs - low_dim_probs, low_dim) * learning_rate
            low_dim += gradients

        # Normalize output
        min_vals = np.min(low_dim, axis=0)
        max_vals = np.max(low_dim, axis=0)
        normalized = (low_dim - min_vals) / (max_vals - min_vals + 1e-6)

        return normalized

    def dbscan(self, points: np.ndarray) -> np.ndarray:
        """Perform DBSCAN clustering."""
        num_points = points.shape[0]
        labels = np.full(num_points, -1)
        
        # Calculate pairwise distances
        distances = np.sqrt(np.sum(points**2, axis=1)[:, np.newaxis] + 
                          np.sum(points**2, axis=1)[np.newaxis, :] - 
                          2 * np.dot(points, points.T))
        
        # Find neighbors
        neighbors = [np.where(distances[i] <= self.eps)[0] for i in range(num_points)]
        
        cluster_id = 0
        for i in range(num_points):
            if labels[i] != -1:
                continue
                
            if len(neighbors[i]) < self.min_pts:
                labels[i] = 0  # Noise
                continue
                
            cluster_id += 1
            labels[i] = cluster_id
            
            # Expand cluster
            seed_set = set(neighbors[i]) - {i}
            while seed_set:
                current = seed_set.pop()
                
                if labels[current] == 0:
                    labels[current] = cluster_id
                    
                if labels[current] != -1:
                    continue
                    
                labels[current] = cluster_id
                
                if len(neighbors[current]) >= self.min_pts:
                    seed_set.update(neighbors[current])
        
        return labels

    def cluster_repositories(self, repos: List[GitHubRepo]) -> List[ClusteredRepo]:
        """Main clustering function."""
        # Extract features
        features = self.extract_features(repos)
        
        # Reduce dimensions and cluster
        try:
            reduced = self.reduce_dimensions(features)
            labels = self.dbscan(reduced)
            
            # Create clustered repositories
            clustered_repos = []
            for i, repo in enumerate(repos):
                clustered_repos.append(ClusteredRepo(
                    repo=repo,
                    cluster_id=int(labels[i]),
                    coordinates=(float(reduced[i, 0]), float(reduced[i, 1]))
                ))
            
            return clustered_repos
            
        except Exception as e:
            print(f"Error in clustering: {e}")
            # Fallback to simple 2D projection
            normalized = (features - np.min(features)) / (np.max(features) - np.min(features) + 1e-6)
            coords = normalized[:, :2]
            
            return [
                ClusteredRepo(
                    repo=repo,
                    cluster_id=0,
                    coordinates=(float(coords[i, 0]), float(coords[i, 1]))
                )
                for i, repo in enumerate(repos)
            ]